src/tool/cpp/helpers.h

```diff
@@ -287,47 +291,56 @@ namespace xlang
 
             info.overridable = overridable || has_attribute(impl, "Windows.Foundation.Metadata", "OverridableAttribute");
             info.base = base;
-            TypeDef definition;
+            info.generic_param_stack = generic_param_stack;
             writer::generic_param_guard guard;
 
-            switch (info.type.type())
+            switch (type.type())
             {
             case TypeDefOrRef::TypeDef:
             {
-                definition = info.type.TypeDef();
+                info.type = type.TypeDef();
                 break;
             }
             case TypeDefOrRef::TypeRef:
             {
-                definition = find_required(info.type.TypeRef());
-                w.add_depends(definition);
+                info.type = find_required(type.TypeRef());
+                w.add_depends(info.type);
                 break;
             }
             case TypeDefOrRef::TypeSpec:
             {
-                auto type_signature = info.type.TypeSpec().Signature();
+                auto type_signature = type.TypeSpec().Signature();
+
+                std::vector<std::string> names;
+
+                for (auto&& arg : type_signature.GenericTypeInst().GenericArgs())
+                {
+                    names.push_back(w.write_temp("%", arg));
+                }
+
+                info.generic_param_stack.push_back(std::move(names));
+
                 guard = w.push_generic_params(type_signature.GenericTypeInst());
                 auto signature = type_signature.GenericTypeInst();
-                definition = find_required(signature.GenericType().TypeRef());
+                info.type = find_required(signature.GenericType().TypeRef());
+
                 break;
             }
             }
 
-            info.methods = definition.MethodList();
-            info.exclusive = has_attribute(definition, "Windows.Foundation.Metadata", "ExclusiveToAttribute");
-            get_interfaces_impl(w, result, info.defaulted, info.overridable, base, definition.InterfaceImpl());
+            get_interfaces_impl(w, result, info.defaulted, info.overridable, base, info.generic_param_stack, info.type.InterfaceImpl());
             result[name] = std::move(info);
         }
     };
 
     auto get_interfaces(writer& w, TypeDef const& type)
     {
         std::map<std::string, interface_info> result;
-        get_interfaces_impl(w, result, false, false, false, type.InterfaceImpl());
+        get_interfaces_impl(w, result, false, false, false, {}, type.InterfaceImpl());
```

Comment:
All these bool flags make the callsites hostile to human readers. Replace the bools with bool enums to preserve readability and prevent accidental mix-and-matching.

Reply:
Agreed, I've been meaning to clean this up as it has evolved into something more complex.